File: renderers/android/src/main/kotlin/dev/varn/gui/VarnCameraView.kt
package dev.varn.gui

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.graphics.SurfaceTexture
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraCaptureSession
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CameraMetadata
import android.hardware.camera2.CaptureRequest
import android.graphics.ImageFormat
import android.graphics.Rect
import android.media.ImageReader
import android.media.MediaRecorder
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import android.util.Size
import android.view.Surface
import android.view.TextureView
import android.view.View
import android.widget.FrameLayout
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID
import kotlin.math.max
import kotlin.math.min

/**
 * What the camera sees, with what it captures written where the tree can show it.
 *
 * The device runs only while the view is attached, so leaving a screen puts the camera down rather than
 * leaving it open behind a screen nobody is looking at. Everything the camera is told is told on a thread
 * of its own, since opening one and configuring a session both block for as long as the hardware takes.
 *
 * A look is drawn over the preview and written into a photograph, which is what makes what a reader sees
 * before they capture the picture they get. A film is recorded as the camera sees it, since the encoder
 * takes frames straight from the hardware and rewriting each one is a different feature.
 */
class VarnCameraView(context: Context) : FrameLayout(context), VarnReleasing {
    /** Told what the camera produced, which is a file the tree can show and what it turned out to be. */
    var onEvent: ((String, JSONObject) -> Unit)? = null

    /** Asked with the permission the camera needs, since only an activity may ask a reader for one. */
    var onPermission: ((String, (Boolean) -> Unit) -> Unit)? = null

    var facing: String = "back"
        set(value) {
            if (field == value) {
                return
            }

            field = value
            reopen()
        }

    var zoom: Float = 1f
        set(value) {
            field = value
            settle()
        }

    var torch: Boolean = false
        set(value) {
            field = value
            settle()
        }

    var wantsAudio: Boolean = false
        set(value) {
            if (field == value) {
                return
            }

            field = value
            reopen()
        }

    var filter: FloatArray? = null
        set(value) {
            field = value
            showFilter()
        }

    private val preview = TextureView(context)

    private val cameras = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
    private var thread: HandlerThread? = null
    private var work: Handler? = null

    private var device: CameraDevice? = null
    private var session: CameraCaptureSession? = null
    private var request: CaptureRequest.Builder? = null
    private var stills: ImageReader? = null
    private var films: MediaRecorder? = null
    private var began: Long = 0

    private var asked = false
    private var opening = false
    private var size = Size(1280, 720)

    init {
        setBackgroundColor(0xff000000.toInt())
        addView(preview, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))

        preview.surfaceTextureListener = object : TextureView.SurfaceTextureListener {
            override fun onSurfaceTextureAvailable(texture: SurfaceTexture, width: Int, height: Int) {
                open()
            }

            override fun onSurfaceTextureSizeChanged(texture: SurfaceTexture, width: Int, height: Int) = Unit

            override fun onSurfaceTextureDestroyed(texture: SurfaceTexture): Boolean {
                close()
                return true
            }

            override fun onSurfaceTextureUpdated(texture: SurfaceTexture) = Unit
        }
    }

    override fun release() {
        close()
    }

    /** Asks the reader for what it needs, and opens the camera once they have answered. */
    private fun open() {
        if (opening || device != null || !preview.isAvailable) {
            return
        }

        opening = true

        allowed(Manifest.permission.CAMERA) { granted ->
            if (!granted) {
                opening = false
                report("onError", JSONObject().put("message", "the camera was refused"))
                return@allowed
            }

            if (!wantsAudio) {
                start()
                return@allowed
            }

            allowed(Manifest.permission.RECORD_AUDIO) { allowedSound ->
                if (!allowedSound) {
                    report("onError", JSONObject().put("message", "the microphone was refused"))
                }

                start()
            }
        }
    }

    /** Asks for a permission once, answering straight away when the reader has already allowed it. */
    private fun allowed(permission: String, then: (Boolean) -> Unit) {
        if (context.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED) {
            then(true)
            return
        }

        val ask = onPermission

        if (ask == null) {
            then(false)
            return
        }

        ask(permission) { granted -> then(granted) }
    }

    private fun start() {
        val id = cameraId()

        if (id == null) {
            opening = false
            report("onError", JSONObject().put("message", "there is no camera to open"))
            return
        }

        val loop = HandlerThread("varn-camera").also { it.start() }

        thread = loop
        work = Handler(loop.looper)
        size = previewSize(id)
        asked = true

        try {
            cameras.openCamera(id, object : CameraDevice.StateCallback() {
                override fun onOpened(opened: CameraDevice) {
                    device = opened
                    opening = false
                    configure()
                }

                override fun onDisconnected(opened: CameraDevice) {
                    opened.close()
                    device = null
                    opening = false
                }

                override fun onError(opened: CameraDevice, problem: Int) {
                    opened.close()
                    device = null
                    opening = false
                    post { report("onError", JSONObject().put("message", "the camera could not be opened")) }
                }
            }, work)
        } catch (problem: CameraAccessException) {
            opening = false
            report("onError", JSONObject().put("message", problem.message ?: "the camera could not be opened"))
        } catch (problem: SecurityException) {
            opening = false
            report("onError", JSONObject().put("message", "the camera was refused"))
        }
    }

    /** Answers the camera facing the way the tree asked for, or whichever one there is. */
    private fun cameraId(): String? {
        val wanted =
            if (facing == "front") CameraCharacteristics.LENS_FACING_FRONT
            else CameraCharacteristics.LENS_FACING_BACK

        val ids = try {
            cameras.cameraIdList
        } catch (problem: CameraAccessException) {
            return null
        }

        for (id in ids) {
            val about = cameras.getCameraCharacteristics(id)

            if (about.get(CameraCharacteristics.LENS_FACING) == wanted) {
                return id
            }
        }

        return ids.firstOrNull()
    }

    /** Answers the size the preview runs at, which is the largest one that is not more than the screen. */
    private fun previewSize(id: String): Size {
        val about = cameras.getCameraCharacteristics(id)
        val map = about.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP) ?: return Size(1280, 720)
        val offered = map.getOutputSizes(SurfaceTexture::class.java) ?: return Size(1280, 720)

        return offered.filter { it.width <= 1920 && it.height <= 1080 }
            .maxByOrNull { it.width.toLong() * it.height } ?: offered.first()
    }

    /** Builds the session the preview and the stills are drawn from, replacing whichever one is running. */
    private fun configure() {
        val opened = device ?: return
        val texture = preview.surfaceTexture ?: return

        texture.setDefaultBufferSize(size.width, size.height)

        val screen = Surface(texture)
        val reader = ImageReader.newInstance(size.width, size.height, ImageFormat.JPEG, 2)

        stills?.close()
        stills = reader

        reader.setOnImageAvailableListener({ from ->
            val image = from.acquireLatestImage() ?: return@setOnImageAvailableListener
            val buffer = image.planes[0].buffer
            val bytes = ByteArray(buffer.remaining())

            buffer.get(bytes)
            image.close()

            post { wrote(bytes) }
        }, work)

        val surfaces = mutableListOf(screen, reader.surface)
        val recorder = films?.surface

        if (recorder != null) {
            surfaces.add(recorder)
        }

        val built = opened.createCaptureRequest(
            if (recorder != null) CameraDevice.TEMPLATE_RECORD else CameraDevice.TEMPLATE_PREVIEW,
        )

        built.addTarget(screen)

        if (recorder != null) {
            built.addTarget(recorder)
        }

        request = built

        @Suppress("DEPRECATION")
        opened.createCaptureSession(surfaces, object : CameraCaptureSession.StateCallback() {
            override fun onConfigured(made: CameraCaptureSession) {
                session = made
                applyRequest()
                post {
                    showFilter()
                    report("onReady", JSONObject().put("facing", facing))
                }
            }

            override fun onConfigureFailed(made: CameraCaptureSession) {
                post { report("onError", JSONObject().put("message", "the camera session could not be built")) }
            }
        }, work)
    }

    /** Puts the camera down, which is what leaving a screen has to do with a device the system lends. */
    private fun close() {
        films?.let { recorder ->
            runCatching { recorder.stop() }
            recorder.release()
        }

        films = null

        session?.close()
        session = null

        device?.close()
        device = null

        stills?.close()
        stills = null

        thread?.quitSafely()
        thread = null
        work = null

        opening = false
    }

    /** Turns the camera round, which is a different device rather than a different setting on this one. */
    private fun reopen() {
        if (!asked) {
            return
        }

        close()
        open()
    }

    /** Applies how far in the camera is and whether its light is on, which the request carries. */
    private fun settle() {
        val built = request ?: return
        val opened = device ?: return

        val about = cameras.getCameraCharacteristics(opened.id)
        val most = about.get(CameraCharacteristics.SCALER_AVAILABLE_MAX_DIGITAL_ZOOM) ?: 1f
        val whole = about.get(CameraCharacteristics.SENSOR_INFO_ACTIVE_ARRAY_SIZE)

        val held = min(max(1f, zoom), most)

        if (whole != null) {
            val width = (whole.width() / held).toInt()
            val height = (whole.height() / held).toInt()
            val left = (whole.width() - width) / 2
            val top = (whole.height() - height) / 2

            built.set(CaptureRequest.SCALER_CROP_REGION, Rect(left, top, left + width, top + height))
        }

        val lit = about.get(CameraCharacteristics.FLASH_INFO_AVAILABLE) == true

        if (lit) {
            built.set(
                CaptureRequest.FLASH_MODE,
                if (torch) CameraMetadata.FLASH_MODE_TORCH else CameraMetadata.FLASH_MODE_OFF,
            )
        }

        applyRequest()
    }

    private fun applyRequest() {
        val made = session ?: return
        val built = request ?: return

        runCatching { made.setRepeatingRequest(built.build(), null, work) }
    }

    /** Draws the preview through the look the tree asked for, which a layer of its own is what applies. */
    private fun showFilter() {
        val matrix = filter

        if (matrix == null || matrix.size != 20) {
            preview.setLayerType(View.LAYER_TYPE_NONE, null)
            return
        }

        val paint = Paint()
        paint.colorFilter = ColorMatrixColorFilter(ColorMatrix(matrix))

        preview.setLayerType(View.LAYER_TYPE_HARDWARE, paint)
    }

    fun capturePhoto() {
        val made = session ?: return
        val opened = device ?: return
        val reader = stills ?: return

        val built = opened.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE)
        built.addTarget(reader.surface)

        request?.get(CaptureRequest.SCALER_CROP_REGION)?.let {
            built.set(CaptureRequest.SCALER_CROP_REGION, it)
        }

        runCatching { made.capture(built.build(), null, work) }
    }

    /** Writes what the hardware produced, drawn through the look the reader was looking at. */
    private fun wrote(bytes: ByteArray) {
        val decoded = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)

        if (decoded == null) {
            report("onError", JSONObject().put("message", "the picture could not be read"))
            return
        }

        val drawn = painted(decoded)
        val file = somewhere("photo", "jpg")

        runCatching {
            file.outputStream().use { out -> drawn.compress(Bitmap.CompressFormat.JPEG, 90, out) }
        }.onFailure { problem ->
            report("onError", JSONObject().put("message", problem.message ?: "the picture could not be written"))
            return
        }

        report(
            "onCapture",
            JSONObject()
                .put("path", file.absolutePath)
                .put("width", drawn.width)
                .put("height", drawn.height),
        )
    }

    /** Answers the picture drawn through the look, which is what makes a capture carry what was seen. */
    private fun painted(source: Bitmap): Bitmap {
        val matrix = filter

        if (matrix == null || matrix.size != 20) {
            return source
        }

        val drawn = Bitmap.createBitmap(source.width, source.height, Bitmap.Config.ARGB_8888)
        val paint = Paint()

        paint.colorFilter = ColorMatrixColorFilter(ColorMatrix(matrix))
        Canvas(drawn).drawBitmap(source, 0f, 0f, paint)

        return drawn
    }

    fun startRecording() {
        if (films != null || device == null) {
            return
        }

        val file = somewhere("film", "mp4")
        val recorder = VarnCameraView.recorder(context)

        runCatching {
            if (wantsAudio) {
                recorder.setAudioSource(MediaRecorder.AudioSource.CAMCORDER)
            }

            recorder.setVideoSource(MediaRecorder.VideoSource.SURFACE)
            recorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            recorder.setOutputFile(file.absolutePath)
            recorder.setVideoEncodingBitRate(6_000_000)
            recorder.setVideoFrameRate(30)
            recorder.setVideoSize(size.width, size.height)
            recorder.setVideoEncoder(MediaRecorder.VideoEncoder.H264)

            if (wantsAudio) {
                recorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            }

            recorder.prepare()
        }.onFailure { problem ->
            recorder.release()
            report("onError", JSONObject().put("message", problem.message ?: "the film could not be started"))
            return
        }

        films = recorder
        recording = file
        began = System.currentTimeMillis()

        // The encoder takes its frames from a surface the session has to carry, so the session is built
        // again with it rather than a recording being started on the one already running.
        session?.close()
        session = null
        configure()

        recorder.start()
    }

    private var recording: File? = null

    fun stopRecording() {
        val recorder = films ?: return
        val file = recording

        films = null
        recording = null

        runCatching { recorder.stop() }
        recorder.release()

        session?.close()
        session = null
        configure()

        if (file == null) {
            return
        }

        report(
            "onRecord",
            JSONObject()
                .put("path", file.absolutePath)
                .put("duration", (System.currentTimeMillis() - began) / 1000.0),
        )
    }

    private fun report(name: String, payload: JSONObject) {
        post { onEvent?.invoke(name, payload) }
    }

    companion object {
        /** Answers where a capture is written, which is a file of the application's own. */
        fun somewhere(context: Context, name: String, ending: String): File {
            return File(context.cacheDir, "varn-$name-${UUID.randomUUID()}.$ending")
        }

        /**
         * Answers a recorder, which from Android 12 is built against the context that owns it.
         *
         * The context is what an attribution is drawn from, so the system can say which part of an
         * application is holding the microphone rather than only that something is.
         */
        fun recorder(context: Context): MediaRecorder {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                return MediaRecorder(context)
            }

            @Suppress("DEPRECATION")
            return MediaRecorder()
        }

        /** Answers the colour matrix a filter arrived as, or nothing when it carries none. */
        fun matrix(value: JSONArray?): FloatArray? {
            if (value == null || value.length() != 20) {
                return null
            }

            return FloatArray(20) {
                val number = value.optDouble(it, 0.0)

                if (it % 5 == 4) (number * 255).toFloat() else number.toFloat()
            }
        }
    }

    private fun somewhere(name: String, ending: String): File = somewhere(context, name, ending)
}

/**
 * A microphone with nothing to draw, which is what a voice note is made of.
 *
 * It is told whether it should be running rather than asked to start, since that is what the tree holds:
 * a state that says recording, and a file that arrives when it stops.
 */
class VarnRecorderView(context: Context) : View(context), VarnReleasing {
    var onEvent: ((String, JSONObject) -> Unit)? = null
    var onPermission: ((String, (Boolean) -> Unit) -> Unit)? = null

    var recording: Boolean = false
        set(value) {
            if (field == value) {
                return
            }

            field = value

            if (value) {
                start()
                return
            }

            finish()
        }

    private var recorder: MediaRecorder? = null
    private var file: File? = null
    private var began: Long = 0

    override fun release() {
        recorder?.let { made ->
            runCatching { made.stop() }
            made.release()
        }

        recorder = null
        file = null
    }

    private fun start() {
        if (context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
            listen()
            return
        }

        val ask = onPermission

        if (ask == null) {
            recording = false
            onEvent?.invoke("onError", JSONObject().put("message", "this application cannot ask to record"))
            return
        }

        ask(Manifest.permission.RECORD_AUDIO) { granted ->
            if (!granted) {
                recording = false
                onEvent?.invoke("onError", JSONObject().put("message", "the microphone was refused"))
                return@ask
            }

            listen()
        }
    }

    private fun listen() {
        val written = VarnCameraView.somewhere(context, "note", "m4a")
        val made = VarnCameraView.recorder(context)

        runCatching {
            made.setAudioSource(MediaRecorder.AudioSource.MIC)
            made.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            made.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            made.setOutputFile(written.absolutePath)
            made.prepare()
            made.start()
        }.onFailure { problem ->
            made.release()
            recording = false
            onEvent?.invoke("onError", JSONObject().put("message", problem.message ?: "the microphone would not open"))
            return
        }

        recorder = made
        file = written
        began = System.currentTimeMillis()

        onEvent?.invoke("onReady", JSONObject())
    }

    private fun finish() {
        val made = recorder ?: return
        val written = file

        recorder = null
        file = null

        runCatching { made.stop() }
        made.release()

        if (written == null) {
            return
        }

        onEvent?.invoke(
            "onFinish",
            JSONObject()
                .put("path", written.absolutePath)
                .put("duration", (System.currentTimeMillis() - began) / 1000.0),
        )
    }
}

File: renderers/ios/VarnCameraView.swift
import UIKit
import AVFoundation

/// What the camera sees, with what it captures written where the tree can show it.
///
/// The session runs only while the view is in a window, so leaving a screen puts the camera down rather
/// than leaving the indicator on over a screen nobody is looking at. Everything the session is told is
/// told on a queue of its own, since configuring one blocks the caller for as long as the hardware takes.
///
/// A look is drawn over the preview and written into a photograph, which is what makes what a reader sees
/// before they capture the picture they get. A film is recorded as the camera sees it, since a platform
/// writes frames straight from the hardware and rewriting each one is a different feature.
final class VarnCameraView: UIView, VarnReleasing {
    /// Told what the camera produced, which is a file the tree can show and what it turned out to be.
    var onEvent: ((String, [String: Any]) -> Void)?

    var facing: String = "back" {
        didSet {
            guard facing != oldValue else {
                return
            }

            frames.async { [weak self, facing] in self?.turned = facing }
            reopen()
        }
    }

    var zoom: CGFloat = 1 {
        didSet { settle() }
    }

    var torch = false {
        didSet { settle() }
    }

    var wantsAudio = false {
        didSet {
            guard wantsAudio != oldValue else {
                return
            }

            reopen()
        }
    }

    var filter: [Double]? {
        didSet {
            frames.async { [weak self] in self?.looked = self?.filter }
            showFilter()
        }
    }

    /// What the frames queue draws through, since the one above it is written from the main thread.
    ///
    /// A sample buffer arrives on a queue of its own and the tree writes the look and which way the
    /// camera is turned on the main one, so reading those from the buffer is two threads on one value.
    private var looked: [Double]?
    private var turned = "back"

    /// Whether a frame is already on its way to the main thread, so the next one is dropped rather
    /// than queued behind it. A camera produces frames faster than a screen can take them, and one
    /// block per frame is a backlog the main thread has to drain before it can do anything else —
    /// which is what leaving the screen had to wait for.
    private var drawing = false

    private let session = AVCaptureSession()
    private let work = DispatchQueue(label: "dev.varn.gui.camera")
    private let frames = DispatchQueue(label: "dev.varn.gui.camera.frames")

    private let photos = AVCapturePhotoOutput()
    private let films = AVCaptureMovieFileOutput()
    private let pictures = AVCaptureVideoDataOutput()

    private let preview = AVCaptureVideoPreviewLayer()

    /// What the filtered frames are drawn into, which stands over the preview while a look is on.
    private let looking = UIImageView()

    private var input: AVCaptureDeviceInput?
    private var opened = false
    private var began: Date?

    private lazy var delegate = VarnCameraDelegate(view: self)

    override init(frame: CGRect) {
        super.init(frame: frame)

        backgroundColor = .black
        clipsToBounds = true

        preview.videoGravity = .resizeAspectFill
        layer.addSublayer(preview)

        looking.contentMode = .scaleAspectFill
        looking.isHidden = true
        looking.isUserInteractionEnabled = false
        addSubview(looking)
    }

    required init?(coder: NSCoder) { fatalError("a camera is built by the factory rather than from a nib") }

    deinit {
        letGo()
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        preview.frame = bounds
        looking.frame = bounds
    }

    /// Runs the camera only while there is a window to draw it in, which is what puts it down on leaving.
    override func didMoveToWindow() {
        super.didMoveToWindow()

        if window == nil {
            work.async { [session] in
                if session.isRunning {
                    session.stopRunning()
                }
            }

            return
        }

        open()
    }

    func letGo() {
        work.async { [session] in
            if session.isRunning {
                session.stopRunning()
            }
        }
    }

    /// Asks the reader for what it needs and opens the camera once they have answered.
    private func open() {
        guard !opened else {
            work.async { [session] in
                if !session.isRunning {
                    session.startRunning()
                }
            }

            return
        }

        opened = true

        VarnCameraView.allowed(for: .video) { [weak self] granted in
            guard let self else {
                return
            }

            guard granted else {
                self.onEvent?("onError", ["message": "the camera was refused"])
                return
            }

            guard self.wantsAudio else {
                self.build()
                return
            }

            VarnCameraView.allowed(for: .audio) { [weak self] allowed in
                guard let self else {
                    return
                }

                if !allowed {
                    self.onEvent?("onError", ["message": "the microphone was refused"])
                }

                self.build()
            }
        }
    }

    /// Asks for a device once, answering straight away when the reader has already been asked.
    private static func allowed(for kind: AVMediaType, then answer: @escaping (Bool) -> Void) {
        switch AVCaptureDevice.authorizationStatus(for: kind) {
        case .authorized:
            answer(true)
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: kind) { granted in
                DispatchQueue.main.async { answer(granted) }
            }
        default:
            answer(false)
        }
    }

    private func build() {
        work.async { [weak self] in
            guard let self else {
                return
            }

            self.session.beginConfiguration()
            self.session.sessionPreset = .high

            self.attachCamera()
            self.attachMicrophone()

            for output in [self.photos, self.films, self.pictures] as [AVCaptureOutput] {
                if self.session.canAddOutput(output) {
                    self.session.addOutput(output)
                }
            }

            self.pictures.alwaysDiscardsLateVideoFrames = true
            self.session.commitConfiguration()

            if !self.session.isRunning {
                self.session.startRunning()
            }

            DispatchQueue.main.async {
                self.preview.session = self.session
                self.settle()
                self.showFilter()
                self.onEvent?("onReady", ["facing": self.facing])
            }
        }
    }

    /// Takes the camera the tree asked for, replacing whichever one the session already had.
    private func attachCamera() {
        let position: AVCaptureDevice.Position = facing == "front" ? .front : .back

        guard let device = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: position),
              let taken = try? AVCaptureDeviceInput(device: device) else {
            DispatchQueue.main.async { self.onEvent?("onError", ["message": "there is no camera to open"]) }
            return
        }

        if let input {
            session.removeInput(input)
        }

        guard session.canAddInput(taken) else {
            DispatchQueue.main.async { self.onEvent?("onError", ["message": "the camera could not be opened"]) }
            return
        }

        session.addInput(taken)
        input = taken
    }

    private func attachMicrophone() {
        guard wantsAudio, let device = AVCaptureDevice.default(for: .audio),
              let taken = try? AVCaptureDeviceInput(device: device), session.canAddInput(taken) else {
            return
        }

        session.addInput(taken)
    }

    /// Turns the camera round, which is a new input on a session that is already running.
    private func reopen() {
        guard opened else {
            return
        }

        work.async { [weak self] in
            guard let self else {
                return
            }

            self.session.beginConfiguration()
            self.attachCamera()
            self.attachMicrophone()
            self.session.commitConfiguration()

            DispatchQueue.main.async {
                self.settle()
                self.onEvent?("onReady", ["facing": self.facing])
            }
        }
    }

    /// Applies how far in the camera is and whether its light is on, which the device itself holds.
    private func settle() {
        guard let device = input?.device else {
            return
        }

        work.async {
            guard (try? device.lockForConfiguration()) != nil else {
                return
            }

            device.videoZoomFactor = min(max(1, self.zoom), device.activeFormat.videoMaxZoomFactor)

            if device.hasTorch && device.isTorchAvailable {
                device.torchMode = self.torch ? .on : .off
            }

            device.unlockForConfiguration()
        }
    }

    /// Shows the filtered frames or the plain preview, which is what having a look or not means.
    /// Shows the look over the preview, and asks for frames only while there is one to draw through.
    ///
    /// A camera hands over every frame it takes, so leaving the delegate attached with no look is thirty
    /// buffers a second crossing a queue to be discarded, and a session that has to drain them before it
    /// can stop.
    private func showFilter() {
        let looking = filter != nil

        self.looking.isHidden = !looking
        preview.isHidden = looking

        if !looking {
            self.looking.image = nil
        }

        pictures.setSampleBufferDelegate(looking ? delegate : nil, queue: looking ? frames : nil)
    }

    /// Draws one frame from the camera through the look the tree asked for.
    fileprivate func draw(_ buffer: CMSampleBuffer) {
        guard let matrix = looked, !drawing, let pixels = CMSampleBufferGetImageBuffer(buffer) else {
            return
        }

        let image = CIImage(cvPixelBuffer: pixels)
        let drawn = VarnFilter.filtered(matrix, image)

        guard let made = VarnFilter.rendered(drawn) else {
            return
        }

        let angled = UIImage(cgImage: made, scale: 1, orientation: turned == "front" ? .leftMirrored : .right)

        drawing = true

        DispatchQueue.main.async { [weak self] in
            guard let self else {
                return
            }

            if self.filter != nil {
                self.looking.image = angled
            }

            self.frames.async { [weak self] in self?.drawing = false }
        }
    }

    /// Takes a picture, which arrives as a file once the hardware and the look are both done with it.
    func capturePhoto() {
        guard session.isRunning else {
            onEvent?("onError", ["message": "the camera is not running"])
            return
        }

        photos.capturePhoto(with: AVCapturePhotoSettings(), delegate: delegate)
    }

    func startRecording() {
        guard session.isRunning, !films.isRecording else {
            return
        }

        began = Date()
        films.startRecording(to: VarnCameraView.somewhere(named: "film", ending: "mov"), recordingDelegate: delegate)
    }

    func stopRecording() {
        guard films.isRecording else {
            return
        }

        films.stopRecording()
    }

    /// Answers where a capture is written, which is a file of the application's own that it may replace.
    static func somewhere(named name: String, ending: String) -> URL {
        let folder = FileManager.default.temporaryDirectory
        let file = folder.appendingPathComponent("varn-\(name)-\(UUID().uuidString).\(ending)")

        return file
    }

    fileprivate func took(photo: Data?, problem: Error?) {
        if let problem {
            onEvent?("onError", ["message": problem.localizedDescription])
            return
        }

        guard let photo, let image = UIImage(data: photo) else {
            onEvent?("onError", ["message": "the picture could not be read"])
            return
        }

        // What a reader was looking at is what they took, so the look is written into the file rather
        // than being something that was only ever on the screen.
        let drawn = VarnFilter.apply(filter, to: image) ?? image

        guard let bytes = drawn.jpegData(compressionQuality: 0.9) else {
            onEvent?("onError", ["message": "the picture could not be written"])
            return
        }

        let file = VarnCameraView.somewhere(named: "photo", ending: "jpg")

        do {
            try bytes.write(to: file)
        } catch {
            onEvent?("onError", ["message": error.localizedDescription])
            return
        }

        onEvent?("onCapture", [
            "path": file.path,
            "width": Double(drawn.size.width * drawn.scale),
            "height": Double(drawn.size.height * drawn.scale),
        ])
    }

    fileprivate func recorded(_ file: URL, problem: Error?) {
        if let problem {
            onEvent?("onError", ["message": problem.localizedDescription])
            return
        }

        let seconds = began.map { Date().timeIntervalSince($0) } ?? 0
        began = nil

        onEvent?("onRecord", ["path": file.path, "duration": seconds])
    }
}

/// What the camera calls back on, kept apart from the view so the view holds no delegate conformances.
private final class VarnCameraDelegate: NSObject, AVCapturePhotoCaptureDelegate,
    AVCaptureFileOutputRecordingDelegate, AVCaptureVideoDataOutputSampleBufferDelegate {
    private weak var view: VarnCameraView?

    init(view: VarnCameraView) {
        self.view = view
    }

    func photoOutput(_ output: AVCapturePhotoOutput, didFinishProcessingPhoto photo: AVCapturePhoto,
                     error: Error?) {
        let data = photo.fileDataRepresentation()

        DispatchQueue.main.async { [weak view] in
            view?.took(photo: data, problem: error)
        }
    }

    func fileOutput(_ output: AVCaptureFileOutput, didFinishRecordingTo file: URL,
                    from connections: [AVCaptureConnection], error: Error?) {
        DispatchQueue.main.async { [weak view] in
            view?.recorded(file, problem: error)
        }
    }

    func captureOutput(_ output: AVCaptureOutput, didOutput buffer: CMSampleBuffer,
                       from connection: AVCaptureConnection) {
        view?.draw(buffer)
    }
}

/// A microphone with nothing to draw, which is what a voice note is made of.
///
/// It is told whether it should be running rather than asked to start, since that is what the tree holds:
/// a state that says recording, and a file that arrives when it stops.
final class VarnRecorderView: UIView, VarnReleasing {
    var onEvent: ((String, [String: Any]) -> Void)?

    var recording = false {
        didSet {
            guard recording != oldValue else {
                return
            }

            if recording {
                start()
                return
            }

            finish()
        }
    }

    private var recorder: AVAudioRecorder?
    private var began: Date?

    func letGo() {
        recorder?.stop()
        recorder = nil
    }

    deinit {
        letGo()
    }

    private func start() {
        AVAudioSession.sharedInstance().requestRecordPermission { [weak self] granted in
            DispatchQueue.main.async {
                guard let self else {
                    return
                }

                guard granted else {
                    self.recording = false
                    self.onEvent?("onError", ["message": "the microphone was refused"])
                    return
                }

                self.listen()
            }
        }
    }

    private func listen() {
        let file = VarnCameraView.somewhere(named: "note", ending: "m4a")

        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100,
            AVNumberOfChannelsKey: 1,
        ]

        do {
            try AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .default)
            try AVAudioSession.sharedInstance().setActive(true)

            let made = try AVAudioRecorder(url: file, settings: settings)
            made.record()

            recorder = made
            began = Date()
            onEvent?("onReady", [:])
        } catch {
            recording = false
            onEvent?("onError", ["message": error.localizedDescription])
        }
    }

    private func finish() {
        guard let recorder else {
            return
        }

        let file = recorder.url
        let seconds = began.map { Date().timeIntervalSince($0) } ?? 0

        recorder.stop()
        self.recorder = nil
        began = nil

        onEvent?("onFinish", ["path": file.path, "duration": seconds])
    }
}
